src/languages/latex.js

```diff
@@ -1,70 +1,231 @@
+  }
+  var ARGUMENT_BRACES = hljs.inherit(
+    BRACE_GROUP_NO_VERBATIM,
+    {endsParent: true, contains: [BRACE_GROUP_NO_VERBATIM, ...EVERYTHING_BUT_BRACES_AND_VERBATIM]}
+  );
+  var ARGUMENT_BRACKETS = {
+    begin: /\[/, end: /\]/,
+    endsParent: true,
+    relevance: 0,
+    contains: [BRACE_GROUP_NO_VERBATIM, ...EVERYTHING_BUT_BRACES_AND_VERBATIM]
+  }
+  var ARGUMENT_ABSENT = {
+    begin: /(?=[.$])/, end: /(?=[.$])/,
+    endsParent: true,
+    relevance: 0
+  }
+  var SPACE_GOBBLER = {
+    begin: /\s+/,
+    endsParent: true,
+    relevance: 0
+  };
+  var ARGUMENT_M = [ARGUMENT_BRACES];
+  var ARGUMENT_O = [ARGUMENT_BRACKETS, ARGUMENT_ABSENT];
+  var ARGUMENT_NONE = [ARGUMENT_ABSENT];
+  var ARGUMENT_AND_THEN = function(arg, starts_mode) {
+    return {
+      contains: [SPACE_GOBBLER],
+      starts: {
         relevance: 0,
-        variants: [
-          {begin: /\$\$/, end: /\$\$/},
-          {begin: /\$/, end: /\$/}
-        ]
+        contains: arg,
+        starts: starts_mode
+      }
+    };
+  };
+  var CSNAME = function(csname, starts_mode) {
+    return hljs.inherit(
+      {
+        begin: '\\\\' + csname + '(?![a-zA-Z@:_])',
+        keywords: {$pattern: /\S+/, keyword: '\\' + csname}
       },
-      hljs.COMMENT(
-        '%',
-        '$',
-        {
-          relevance: 0
-        }
-      )
+      ARGUMENT_AND_THEN(ARGUMENT_NONE, starts_mode)
+    );
+  };
+  var BEGIN_ENV = function(envname, starts_mode) {
+    return hljs.inherit(
+      {
+        begin: '\\\\begin(?=\\s*\\r?\\n?\\s*\\{' + envname + '\\})',
+        keywords: {$pattern: /\S+/, keyword: '\\begin'}
+      },
+      ARGUMENT_AND_THEN(ARGUMENT_M, starts_mode)
+    );
+  };
+  var VERBATIM_DELIMITED_EQUAL = hljs.END_SAME_AS_BEGIN({
+    className: 'string',
+    begin: /(.|\r?\n)/, end: /(.|\r?\n)/,
+    excludeBegin: true, excludeEnd: true,
+    endsParent: true
+  });
+  var VERBATIM_DELIMITED_BRACES = {
+    className: 'string',
+    begin: /\{/, end: /\}/,
+    excludeBegin: true, excludeEnd: true,
+    endsParent: true
+  };
+  var VERBATIM_DELIMITED_ENV = function(envname) {
+    return {
+      className: 'string',
+      end: '(?=\\\\end\\{' + envname + '\\})'
+    };
+  };
+  var PATCH_CLASSNAME = function(mode, class_name) {
+    return hljs.inherit(mode, {className: class_name});
+  };
+  var VERBATIM = {
+    variants: [
+      CSNAME('verb', {contains: [VERBATIM_DELIMITED_EQUAL]}),
+      CSNAME('lstinline', {contains: [VERBATIM_DELIMITED_EQUAL]}),
+      CSNAME('mint', ARGUMENT_AND_THEN(ARGUMENT_M, {contains: [VERBATIM_DELIMITED_EQUAL]})),
+      CSNAME('mintinline', ARGUMENT_AND_THEN(ARGUMENT_M, {contains: [VERBATIM_DELIMITED_BRACES, VERBATIM_DELIMITED_EQUAL]})),
+      CSNAME('url', {contains: [PATCH_CLASSNAME(VERBATIM_DELIMITED_BRACES, 'link'),
+                                PATCH_CLASSNAME(VERBATIM_DELIMITED_EQUAL, 'link')]}),
+      BEGIN_ENV('verbatim',    {starts: VERBATIM_DELIMITED_ENV('verbatim')}),
```

Comment:
What are all these "*" variants?  Looks like we could do this with a loop, no?  Defining both rules from a single list? (other than the minted exception)

Reply:
Yes, we could also combine the `[BL]Verbatim\*?` variants into a loop. Not sure about the clearest way to do that, though. Do you mean just adding a bare loop that modifies the `VERBATIM` object?

Reply:
```
// for
      BEGIN_ENV(X,     ARGUMENT_AND_THEN(ARGUMENT_O, {starts: VERBATIM_DELIMITED_ENV('Verbatim')})),
      BEGIN_ENV(X + "*",  ARGUMENT_AND_THEN(ARGUMENT_O, {starts: VERBATIM_DELIMITED_ENV('Verbatim\\*')})),
// end
```

Reply:
Are you writing pseudocode with those comments or is there some preprocessor magic I'm not aware of? (If so, what should I google?)

Reply:
Could you clarify this for me? I'm still a bit unsure how to interpret this code snippet. I would have maybe tried something like
```
...['verbatim', 'verbatim\\*'].map(csname => BEGIN_ENV(csname, VERBATIM_DELIMITED_ENV(csname))),
```
but it seems like you had a different (possibly clearer) syntax in mind.